lib/Sema/TypeCheckDistributed.cpp

```diff
@@ -127,44 +149,95 @@ bool swift::checkDistributedFunction(FuncDecl *func, bool diagnose) {
+  if (!swift::ensureDistributedModuleLoaded(ctor))
     return;
+
+  // === Designated initializers must accept exactly one ActorTransport
+  auto &C = ctor->getASTContext();
+  auto module = ctor->getParentModule();
+
+  SmallVector<ParamDecl*, 2> transportParams;
+  int transportParamsCount = 0;
+  auto protocolDecl = C.getProtocol(KnownProtocolKind::ActorTransport);
+  auto protocolTy = protocolDecl->getDeclaredInterfaceType();
+
+  for (auto param : *ctor->getParameters()) {
+    auto paramTy = ctor->mapTypeIntoContext(param->getInterfaceType());
+    auto conformance = TypeChecker::conformsToProtocol(paramTy, protocolDecl, module);
+
+    if (paramTy->isEqual(protocolTy) || !conformance.isInvalid()) {
+      transportParamsCount += 1;
+      transportParams.push_back(param);
+    }
   }
 
-  if (ctor->isDistributedActorResolveInit()) {
-    // It is illegal for users to attempt defining a resolve initializer;
-    // Suggest removing it entirely, there is no way users can implement this init.
-    ctor->diagnose(diag::distributed_actor_init_resolve_must_not_be_user_defined)
-        .fixItRemove(SourceRange(ctor->getStartLoc(), decl->getStartLoc()));
+  // missing transport parameter
+  if (transportParamsCount == 0) {
+    ctor->diagnose(diag::distributed_actor_designated_ctor_missing_transport_param,
+                   ctor->getName());
+    // TODO(distributed): offer fixit to insert 'transport: ActorTransport'
     return;
   }
 
-  // All user defined initializers on distributed actors must be 'convenience'.
-  //
-  // The only initializer that is allowed to be designated is init(transport:)
-  // which we synthesize on behalf of a distributed actor.
-  //
-  // When checking ctor bodies we'll check
-  if (!ctor->isConvenienceInit()) {
-    ctor->diagnose(diag::distributed_actor_init_user_defined_must_be_convenience,
-                   ctor->getName())
-        .fixItInsert(ctor->getConstructorLoc(), "convenience ");
+  // ok! We found exactly one transport parameter
+  if (transportParamsCount == 1)
     return;
-  }
+
+  // TODO(distributed): could list exact parameters
+  ctor->diagnose(diag::distributed_actor_designated_ctor_must_have_one_transport_param,
+                 ctor->getName(), transportParamsCount);
+}
+
+// ==== ------------------------------------------------------------------------
+
+void TypeChecker::checkDistributedActor(ClassDecl *decl) {
+  // ==== Ensure the _Distributed module is available,
+  // without it there's no reason to check the decl in more detail anyway.
+  if (!swift::ensureDistributedModuleLoaded(decl))
+    return;
+
+  // ==== Constructors
+  // --- Get the default initializer
+  // If applicable, this will the default 'init(transport:)' initializer
+  (void)decl->getDefaultInitializer();
+
+  // --- Check all constructors
+  for (auto member : decl->getMembers())
+    if (auto ctor = dyn_cast<ConstructorDecl>(member))
+      checkDistributedActorConstructor(decl, ctor);
```

Comment:
Since we have this, do we really need `NominalTypeDecl::hasUserDefinedDesignatedInit` anymore? I couldn't spot any uses of that in this PR, but maybe I missed it.

Reply:
That one was tricky, thanks for spotting it -- I agree we can remove it from NominalTypeDecl.

So the request always existed, but is only exposed as a static func, no idea why not on the decl itself.
I've removed what we added (on the nominal) since indeed we don't need it since we rely on the default constructor to create "our" default init.